### PR TITLE
Use channel in messaging between goroutine in stead of zmq.REQ

### PR DESCRIPTION
**Pull request by rickmak**
This will resolve socket exhaust problem on high concurrency.

connects #160

### CODE REVIEW

**Comment by cheungpat** [comment]
timeout when running the test with the `./pkg/server/plugin/zmq` package (used `-timeout 5m`)

i hope these are helpful:

```
time="2016-10-22T09:08:52Z" level=info msg="zmq/broker: disconnected worker = address1" logger=plugin
time="2016-10-22T09:08:52Z" level=info msg="zmq/broker: disconnected worker = address2" logger=plugin
.
16 total assertions

time="2016-10-22T09:08:52Z" level=info msg="zmq channler running 0xc820120080\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:08:52Z" level=info msg="zmq/broker: ready worker = ready" logger=plugin
.time="2016-10-22T09:08:53Z" level=info msg="zmq/broker: shutdown of worker = ready" logger=plugin
time="2016-10-22T09:08:53Z" level=info msg="zmq/broker: ready worker = ready1" logger=plugin
time="2016-10-22T09:08:53Z" level=info msg="zmq/broker: ready worker = ready2" logger=plugin
time="2016-10-22T09:08:53Z" level=warning msg="zmq/broker: Ticking non-registered worker = ready" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:08:53Z" level=info msg="zmq/broker: shutdown of worker = ready" logger=plugin
.time="2016-10-22T09:08:54Z" level=info msg="zmq/broker: shutdown of worker = ready1" logger=plugin
time="2016-10-22T09:08:54Z" level=info msg="zmq/broker: shutdown of worker = ready2" logger=plugin
time="2016-10-22T09:08:54Z" level=warning msg="zmq/broker: Ticking non-registered worker = heartbeat" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:08:54Z" level=warning msg="zmq/broker: Ticking non-registered worker = ready1" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:08:54Z" level=info msg="zmq/broker: shutdown of worker = ready1" logger=plugin
time="2016-10-22T09:08:54Z" level=warning msg="zmq/broker: Ticking non-registered worker = ready2" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:08:54Z" level=info msg="zmq/broker: shutdown of worker = ready2" logger=plugin
.time="2016-10-22T09:08:55Z" level=warning msg="zmq/broker: Ticking non-registered worker = heartbeat" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:08:55Z" level=info msg="zmq/broker: shutdown of worker = heartbeat" logger=plugin
time="2016-10-22T09:08:55Z" level=warning msg="zmq/broker: Ticking non-registered worker = unregistered" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:08:55Z" level=info msg="zmq/broker: chan not found for worker unregistered\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
.time="2016-10-22T09:08:56Z" level=warning msg="zmq/broker: Ticking non-registered worker = unregistered" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:08:56Z" level=info msg="zmq/broker: shutdown of worker = unregistered" logger=plugin
time="2016-10-22T09:08:56Z" level=info msg="zmq/broker: ready worker = timeout" logger=plugin
.time="2016-10-22T09:09:00Z" level=info msg="zmq/broker: chan time out for  worker timeout\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
.time="2016-10-22T09:09:00Z" level=info msg="zmq/broker: shutdown of worker = timeout" logger=plugin
time="2016-10-22T09:09:00Z" level=info msg="zmq/broker: no worker avaliable, retry 1...\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:01Z" level=info msg="zmq/broker: no worker avaliable, retry 2...\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:02Z" level=info msg="zmq/broker: no worker avaliable, retry 3...\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:03Z" level=info msg="zmq/broker: no worker avaliable, timeout.\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
..time="2016-10-22T09:09:03Z" level=info msg="zmq/broker: no worker avaliable, retry 1...\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:04Z" level=info msg="zmq/broker: no worker avaliable, retry 2...\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:04Z" level=info msg="zmq/broker: ready worker = lateworker" logger=plugin
...time="2016-10-22T09:09:05Z" level=info msg="zmq/broker: shutdown of worker = lateworker" logger=plugin
time="2016-10-22T09:09:05Z" level=info msg="zmq/broker: ready worker = worker" logger=plugin
time="2016-10-22T09:09:08Z" level=info msg="zmq/broker: chan time out for  worker lateworker\n" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:08Z" level=info msg="zmq/broker: disconnected worker = worker" logger=plugin
xtime="2016-10-22T09:09:10Z" level=warning msg="zmq/broker: Ticking non-registered worker = worker" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:10Z" level=info msg="zmq/broker: shutdown of worker = worker" logger=plugin
time="2016-10-22T09:09:10Z" level=info msg="zmq/broker: ready worker = worker1" logger=plugin
time="2016-10-22T09:09:10Z" level=info msg="zmq/broker: ready worker = worker2" logger=plugin
time="2016-10-22T09:09:13Z" level=info msg="zmq/broker: disconnected worker = worker1" logger=plugin
time="2016-10-22T09:09:13Z" level=info msg="zmq/broker: disconnected worker = worker2" logger=plugin
time="2016-10-22T09:09:15Z" level=warning msg="zmq/broker: Ticking non-registered worker = worker2" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:15Z" level=info msg="zmq/broker: shutdown of worker = worker2" logger=plugin
time="2016-10-22T09:09:15Z" level=warning msg="zmq/broker: Ticking non-registered worker = worker1" eaddr="inproc://plugin.test" logger=plugin plugin=test
time="2016-10-22T09:09:15Z" level=info msg="zmq/broker: shutdown of worker = worker1" logger=plugin
panic: test timed out after 5m0s

goroutine 35 [running]:
panic(0x7a17c0, 0xc820138f10)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:725 +0x14b
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x3a

goroutine 1 [chan receive]:
testing.RunTests(0x9b2d60, 0xd21c40, 0x2, 0x2, 0xc82007ab01)
	/usr/local/go/src/testing/testing.go:583 +0x8d2
testing.(*M).Run(0xc820047ef8, 0xd3fc00)
	/usr/local/go/src/testing/testing.go:515 +0x81
main.main()
	github.com/skygeario/skygear-server/pkg/server/plugin/zmq/_test/_testmain.go:108 +0x253

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1

goroutine 18 [chan receive]:
github.com/skygeario/skygear-server/pkg/server/plugin/zmq.TestBrokerWorker.func1.9(0x7f0e120efc70, 0xc82000eb40)
	/go/src/github.com/skygeario/skygear-server/pkg/server/plugin/zmq/broker_test.go:314 +0x1a5
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.(*context).conveyInner(0xc82000eb40, 0x977620, 0x2b, 0xc820165800)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/context.go:261 +0x21a
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.(*context).Convey.func1()
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/context.go:163 +0x40
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls._m(0x0, 0xc820165820)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/stack_tags.go:39 +0x2b
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls.mark1(0x0, 0xc820165820)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/stack_tags.go:18 +0x2b
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls._m(0x1, 0xc820165820)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/stack_tags.go:41 +0x5a
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls.markS(0x1, 0xc820165820)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/stack_tags.go:16 +0x2b
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls.addStackTag(0x1, 0xc820165820)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/stack_tags.go:13 +0x37
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls.(*ContextManager).SetValues(0xc8200bb200, 0xc82016ce70, 0xc820165820)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/context.go:92 +0x4b3
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.(*context).Convey(0xc82010c2a0, 0xc8200499c0, 0x2, 0x2)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/context.go:164 +0x36b
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.Convey(0xc8200499c0, 0x2, 0x2)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/doc.go:77 +0x62
github.com/skygeario/skygear-server/pkg/server/plugin/zmq.TestBrokerWorker.func1()
	/go/src/github.com/skygeario/skygear-server/pkg/server/plugin/zmq/broker_test.go:316 +0xd8b
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.parseAction.func1(0x7f0e120efc70, 0xc82010c2a0)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/discovery.go:80 +0x18
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.(*context).conveyInner(0xc82010c2a0, 0x8ed6f0, 0xb, 0xc820106390)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/context.go:261 +0x21a
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.rootConvey.func1()
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/context.go:110 +0x142
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls._m(0x0, 0xc82010e280)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/stack_tags.go:39 +0x2b
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls.markS(0x0, 0xc82010e280)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/stack_tags.go:16 +0x2b
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls.addStackTag(0x0, 0xc82010e280)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/stack_tags.go:13 +0x37
github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls.(*ContextManager).SetValues(0xc8200bb200, 0xc820114270, 0xc82010e280)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/jtolds/gls/context.go:92 +0x4b3
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.rootConvey(0xc820049f20, 0x3, 0x3)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/context.go:113 +0x3eb
github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey.Convey(0xc820049f20, 0x3, 0x3)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/smartystreets/goconvey/convey/doc.go:75 +0x45
github.com/skygeario/skygear-server/pkg/server/plugin/zmq.TestBrokerWorker(0xc820124000)
	/go/src/github.com/skygeario/skygear-server/pkg/server/plugin/zmq/broker_test.go:372 +0x337
testing.tRunner(0xc820124000, 0xd21c58)
	/usr/local/go/src/testing/testing.go:473 +0x98
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:582 +0x892

goroutine 19 [syscall, locked to thread]:
github.com/skygeario/skygear-server/vendor/github.com/zeromq/goczmq._Cfunc_zpoller_wait(0x7f0e00009780, 0x7f0e000003e8, 0x0)
	github.com/skygeario/skygear-server/vendor/github.com/zeromq/goczmq/_obj/_cgo_gotypes.go:628 +0x42
github.com/skygeario/skygear-server/vendor/github.com/zeromq/goczmq.(*Poller).Wait(0xc82010e3e0, 0x3e8, 0xd3fcc0)
	/go/src/github.com/skygeario/skygear-server/vendor/github.com/zeromq/goczmq/poller.go:87 +0xd7
github.com/skygeario/skygear-server/pkg/server/plugin/zmq.(*Broker).Run(0xc820120080)
	github.com/skygeario/skygear-server/pkg/server/plugin/zmq/_test/_obj_test/broker.go:142 +0x4c2
created by github.com/skygeario/skygear-server/pkg/server/plugin/zmq.NewBroker
	github.com/skygeario/skygear-server/pkg/server/plugin/zmq/_test/_obj_test/broker.go:108 +0x5e0

goroutine 20 [chan send]:
github.com/skygeario/skygear-server/pkg/server/plugin/zmq.(*Broker).Channeler(0xc820120080)
	github.com/skygeario/skygear-server/pkg/server/plugin/zmq/_test/_obj_test/broker.go:289 +0x1251
created by github.com/skygeario/skygear-server/pkg/server/plugin/zmq.NewBroker
	github.com/skygeario/skygear-server/pkg/server/plugin/zmq/_test/_obj_test/broker.go:109 +0x602

goroutine 28 [chan send]:
github.com/skygeario/skygear-server/pkg/server/plugin/zmq.(*Broker).setTimeout(0xc820120080, 0xc82015f170, 0xa, 0xb2d05e00)
	github.com/skygeario/skygear-server/pkg/server/plugin/zmq/_test/_obj_test/broker.go:312 +0x67
created by github.com/skygeario/skygear-server/pkg/server/plugin/zmq.(*Broker).Channeler
	github.com/skygeario/skygear-server/pkg/server/plugin/zmq/_test/_obj_test/broker.go:278 +0xf19
FAIL	github.com/skygeario/skygear-server/pkg/server/plugin/zmq	300.016s
```

File: pkg/server/plugin/zmq/broker.go [hunk 1]
```diff
@@ -135,110 +150,223 @@ func (lb *Broker) Run() {
...
+			frames, err := pull.RecvMessage()
 			if err != nil {
 				panic(err)
 			}
-
-			frames = append([][]byte{lb.workers.Next()}, frames...)
-			lb.backend.SendMessage(frames)
-			lb.logger.Debugf("zmq/broker: server => plugin: %#x, %s\n", frames[0], frames)
+			backend.SendMessage(frames)
 		case nil:
```
Comment (cheungpat) [inline pkg/server/plugin/zmq/broker.go]: you can also quit the loop by checking the `stop` channel here

File: pkg/server/plugin/zmq/broker.go [hunk 2]
```diff
@@ -50,81 +50,96 @@ const (
...
-	logger        *logrus.Entry
+	name        string
+	backendAddr string
+	frontend    chan [][]byte
+	recvChan    chan *parcel
+	addressChan map[string]chan []byte
+	timeout     chan string
+	workers     workerQueue
+	logger      *logrus.Entry
+	stop        chan int
```
Comment (cheungpat) [inline pkg/server/plugin/zmq/broker.go]: while you still have fresh memory what this struct does, could you add some comments to the meaning of the struct members?

Comment (cheungpat) [inline pkg/server/plugin/zmq/broker.go]: if you have still got time: there are two components to this broker: one is a broker and the other is the channeler. There are some struct members that are only used in one component but not the other, perhaps we could make the distinction more obvious by creating separate structs for these two components.

**Comment by cheungpat** [comment]
if you enable edits on the branch for the pull request, I could make assorted changes such as typos on the pull request on your behalf

File: pkg/server/plugin/zmq/broker.go [hunk 3]
```diff
@@ -135,110 +150,223 @@ func (lb *Broker) Run() {
...
+//
+// Worker is expect to register itself on ready. Tick itself when it is
+// avalible. The most recently Tick worker will got the job.
+// A worker do not Tick itself within the expiry will regards as disconnected
+// and requires to Add itself again to become avaliable.
+//
+// workerQueue is not goroutine safe. To use it safely across goroutine.
+// Please use the Lock/Unlock interace before manupliate the queue item via
+// methods like Add/Tick/Purge.
+// Comsuming method Next is the only method will acquire the mutex lock
```
Comment (cheungpat) [inline pkg/server/plugin/zmq/broker.go]: “Comsuming”?

File: pkg/server/plugin/zmq/broker.go (hunk 1 above)
Comment (rickmak) [inline pkg/server/plugin/zmq/broker.go]: No. This switch case is only zmq socket. Not go chan. So we cannot consume the chan here.

**Comment by rickmak** [comment]
The timeout is because Travis running the test with `GOMAXPROCS=1`. When we call zmq code, the runtime internally calls LockOSThread since it is C code. So it result in a dead lock.

Setting `GOMAXPROCS=2` (or larger number) will make the test pass.

**Comment by rickmak** [comment]
Making another PR.

**Pull request closed without merging**